specifyweb/frontend/js_src/lib/components/Forms/ResourceView.tsx

```diff
@@ -182,6 +184,26 @@ export function ResourceView<SCHEMA extends AnySchema>({
       </Dialog>
     );
 
+  const [recordInReadOnlyPref] = userPreferences.use(
+    'form',
+    'recordSet',
+    'isReadOnly'
+  );
+
+  const [recordInReadOnly, setRecordInReadOnly] =
+    React.useState(recordInReadOnlyPref);
+
+  React.useEffect(() => {
+    if (isRecordInReadOnly !== undefined)
+      setRecordInReadOnly(isRecordInReadOnly);
+  }, [isRecordInReadOnly]);
+
+  const editRecord = recordInReadOnly ? (
+    <Button.Secondary onClick={() => setRecordInReadOnly(false)}>
+      {commonText.edit()}
+    </Button.Secondary>
+  ) : undefined;
```

Comment:
You shouldn't need to add any new props at all. `isRecordInReadOnly` shouldn't be needed at all

Just consider this code:

```tsx
const isReadOnly = React.useContext(ReadOnlyContext);
const [temporaryReadOnly, _,_, handleToggleReadOnly] = useBooleanState(recordInReadOnly);
const button = temporaryReadOnly && !isReadOnly
   ? <Button.Secondary onClick={handleToggleReadOnly}>{commonText.edit()}</Button.Secondary>
   : saveButtonElement;
```

And then:

```tsx
<ReadOnlyContext.Provider value={temporaryReadOnly || isReadOnly}>
  {/* the form and rest of the stuff goes here */}
</ReadOnlyContext.Provider>
```

Reply:
The issue with this is that it doesn't handle all the cases. By example when we do a new data entry it uses ResourceView as well now. In the attachment gallery I had to add some logic too. 
I tested this branch and it seems to work everywhere and to handle all the cases. 
I'm happy to discuss this more over a quick call if you have a few minutes at some point

Reply:
I'll try that

Reply:
Do we want to have a single edit button that controls everything or separate edit buttons for the main form and the sub view dialogs/attachment dialogs?
If we only need a single edit button, then there is no need for a prop. If you want to be able to independently enable read-only mode for the main form, but not the sub form, then you likely need a prop.

Reply:
I was thinking having it independent. I like to be able to change the readonly default behavior with this prop

Reply:
We really only need a single edit button on the main form (just like in Specify 6). It doesn't make sense to repeat it, and users would understand that they need to close the dialog to enable edit more.

Reply:
Sounds good, because a single edit button will simplify the code. @CarolineDenis is there a problem with just having an edit button in place of the main save button?

Reply:
There is only a single button
![Screenshot 2023-06-06 at 7 46 04 AM](https://github.com/specify/specify7/assets/108160931/5038a430-30a7-4ba7-b625-53b73845de8c)
![Screenshot 2023-06-06 at 7 46 41 AM](https://github.com/specify/specify7/assets/108160931/680e2551-11e9-4f34-9ee1-bd85107aea2e)

Reply:
yes, but if you open a subview in a dialog, or an attachment viewer, you have another edit button
if I read your code correctly, you can set subview to read only or not independently of the main form, which might be not a useful feature and just add complexity for the user and in the code

Reply:
In a subview or attachment viewer I did it like this: 
![Screenshot 2023-06-06 at 8 37 38 AM](https://github.com/specify/specify7/assets/108160931/d732d750-d562-47d4-b982-b9893b6e7ae5)

Reply:
I was thinking we could simpliy the code by removing the edit button from the attachments viewer and just have it listen to ReadOnlyContext
Then, whether you are in read only or not is controlled only by a single "Edit" button on the **main** specify form (which would be behind the dialog)

Reply:
But that wouldn't work in the attachment gallery as there is no main form behind it

Reply:
Or we say that the attachment form in the plugin opened from the gallery is not readonly from start

Reply:
> But that wouldn't work in the attachment gallery as there is no main form behind it

Attachment gallery is used in two places if I remember correctly:
1. The form, in which case the ResourceView from the main form will call ReadOnlyContext with value true, so the attachment gallery will be read only
2. The attachment viewer page, which is less commonly used than the main form so if it does or doesn't have an edit button is not as much of a deal. Still, the ResoureView is called by Attachment gallery, so it might end up having an edit button anyway.

> Or we say that the attachment form in the plugin opened from the gallery is not readonly from start

Sorry, don't quite understand it.
When user opens let's say Collection Object, it is read only, and everything inside of it, including Attachment Viewer is read only.
When user clicks on "Edit" on the Collection Object form, everything inside of it becomes non-read only, including the ResourceView that is called inside of Attachment Viewer
No special behaviour is needed for attachment viewer at all, or am I missing something?

This PR should probably only need to modify the ResourceView file, with just a few line changes, and nothing else

Reply:
If we click on an attachment in the AttachmentGallery it opens attachment preview with attachment form. If the user choose to have his recordSet readOnly by default the attachment form in the dialog is not editable. The only way to edit it is to go in pref than choose not read only by default than go back to attachments. Which is why I added a way to also edit this form in Viewer.tsx
![Screenshot 2023-06-09 at 9 02 14 AM](https://github.com/specify/specify7/assets/108160931/2ab09bcd-f09b-4949-b0aa-ac3e49a2c087)

Reply:
See #3612
I believe in that implementation you can't make attachment viewer in the attachments tool read-only, but most importantly everything on the main form works, and the code is way simpler